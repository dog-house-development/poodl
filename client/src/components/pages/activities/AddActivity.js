import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ActivityActions from '../../../actions/activityActions';
import _ from 'lodash';
import moment from 'moment';
import DatePicker from '../../ui/DatePicker';
import TimePicker from '../../ui/TimePicker';
import Field from '../../ui/Field';
import Button from '../../ui/Button';
import ComboBox from '../../ui/ComboBox';

const propTypes = {
    auth: PropTypes.object.isRequired,
    errors: PropTypes.oneOfType([PropTypes.object, PropTypes.string])
};

export class AddActivity extends Component {
    constructor(props) {
        super(props);
        this.state = {
            seniorCenterId: props.adminSeniorCenterId,
            multiDay: false,
            startDate: moment().startOf('day'),
            endDate: moment()
                .startOf('day')
                .add(1, 'day'),
            name: '',
            startTime: moment(moment().get('hours'), 'h')
        };

        this.state.endTime = this.state.startTime.clone().add(1, 'hour');
    }

    componentDidMount() {
        this.props.activityActions.filter();
    }

    onValueChange = (name, value) => {
        this.setState({ [name]: value });
    };

    onChange = e => {
        this.setState({ [e.target.id]: e.target.value });
    };

    onSubmit = e => {
        e.preventDefault();
        const startTimeDuration = moment.duration(this.state.startTime.diff(moment().startOf('day')));
        const endTimeDuration = moment.duration(this.state.endTime.diff(moment().startOf('day')));
        const start = this.state.startDate.clone().add(startTimeDuration);
        let end = this.state.startDate.clone().add(endTimeDuration);

        if (this.state.multiDay) {
            end = this.state.endDate.clone().add(endTimeDuration);
        }

        const newActivity = {
            ...this.state,
            startDate: start.toISOString(),
            endDate: end.toISOString()
        };
        console.log(newActivity);
        this.props.activityActions.create(newActivity, this.props.history);
    };

    getFields = () => {
        const fields = [
            {
                onChange: this.onChange,
                error: this.props.errors.description,
                id: 'description',
                type: 'text',
                label: 'Description',
                placeholder: 'Description',
                autoComplete: 'off'
            }
        ];

        return fields;
    };

    getFieldsMarkup() {
        return _.map(this.getFields(), field => <Field key={field.id} {...field} />);
    }

    getDateErrorMarkup() {
        if (this.props.errors.startDate) {
            if (this.state.multiDay) {
                return this.props.errors.startDate;
            }

            return 'Start time must be before end time';
        }
    }

    toggleMultiDay = () => {
        this.setState(state => {
            return { multiDay: !state.multiDay };
        });
    };

    getEndDateMarkup() {
        if (this.state.multiDay) {
            return (
                <DatePicker title="End Date" name="endDate" date={this.state.endDate} onChange={this.onValueChange} />
            );
        }
    }

    getDatePickersMarkup() {
        return (
            <>
                <div className="side-by-side">
                    <DatePicker
                        title={this.state.multiDay ? 'Start Date' : 'Date'}
                        name="startDate"
                        date={this.state.startDate}
                        onChange={this.onValueChange}
                        error={this.state.multiDay ? this.getDateErrorMarkup() : ''}
                    />
                    {this.getEndDateMarkup()}
                </div>
                <Button
                    type="button"
                    size="small"
                    kind="secondary"
                    content={this.state.multiDay ? 'One Day' : 'Multiple Days'}
                    onClick={this.toggleMultiDay}
                    title={
                        this.state.multiDay
                            ? 'The activity occurs during a single day'
                            : 'The activity spans multiple days'
                    }
                />
            </>
        );
    }

    render() {
        return (
            <div className="add-activity-container">
                <Link to="/activities" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to all activities
                </Link>
                <div className="add-activity-panel panel">
                    <h1 className="panel-title">Add Activity</h1>
                    <form noValidate onSubmit={this.onSubmit}>
                        <ComboBox
                            data={this.props.activities}
                            label="Name"
                            placeholder="Name"
                            id="name"
                            value={this.state.name}
                            onChange={this.onChange}
                            error={this.props.errors.name}
                        />
                        {this.getFieldsMarkup()}
                        <div className="side-by-side">
                            <TimePicker
                                title="Start Time"
                                name="startTime"
                                time={this.state.startTime}
                                onChange={this.onValueChange}
                                error={this.state.multiDay ? '' : this.getDateErrorMarkup()}
                            />
                            <TimePicker
                                title="End Time"
                                name="endTime"
                                time={this.state.endTime}
                                onChange={this.onValueChange}
                            />
                        </div>
                        {this.getDatePickersMarkup()}
                        <Button type="submit" content="Add Activity" formButton />
                    </form>
                </div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        auth: state.auth,
        adminIsSuper: _.get(state.auth.admin, 'superAdmin', false),
        adminSeniorCenterId: _.get(state.auth.admin, 'seniorCenterId'),
        errors: state.activities.errors,
        activities: _.get(state.activities, 'all')
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        activityActions: bindActionCreators(ActivityActions, dispatch)
    };
};

AddActivity.propTypes = propTypes;
export default connect(
    mapStateToProps,
    mapDispatchToProps
)(AddActivity);
