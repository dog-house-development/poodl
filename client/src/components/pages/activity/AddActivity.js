import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addActivity } from '../../../actions/activityActions';
import _ from 'lodash';

import Form from '../../ui/Form';

const propTypes = {
    registerActivity: PropTypes.func.isRequired,
    auth: PropTypes.object.isRequired,
    errors: PropTypes.object.isRequired
};

export class AddActivity extends Component {
    constructor(props) {
        super(props);
        this.state = {
            seniorCenter: props.adminSeniorCenter,
            errors: {}
        };
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.errors) {
            this.setState({
                errors: nextProps.errors
            });
        }
    }

    onChange = e => {
        this.setState({ [e.target.id]: e.target.value });
    };

    onSubmit = e => {
        e.preventDefault();
        const newActivity = {
            ...this.state
        };

        this.props.registerActivity(newActivity, this.props.history);
    };

    getFields = () => {
        const { errors } = this.state;
        const fields = [
            {
                onChange: this.onChange,
                error: errors.name,
                id: 'name',
                type: 'text',
                label: 'Name',
                placeholder: 'Name',
                autoComplete: 'off'
            },
            {
                onChange: this.onChange,
                error: errors.description,
                id: 'description',
                type: 'text',
                label: 'Description',
                placeholder: 'Description',
                autoComplete: 'off'
            },
            {
                onChange: this.onChange,
                error: errors.startDate,
                id: 'startDate',
                type: 'datetime-local',
                label: 'Start date',
                sidebyside: 1
            },
            {
                onChange: this.onChange,
                error: errors.endDate,
                id: 'endDate',
                type: 'datetime-local',
                label: 'End Date',
                sidebyside: 2
            }
        ];

        return fields;
    };

    render() {
        return (
            <div className="activity-container">
                <Link to="/activities" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to all activities
                </Link>
                <Form
                    noValidate
                    fields={this.getFields()}
                    onSubmit={this.onSubmit}
                    buttonLabel="Add Activity"
                    formTitle="Add Activity"
                />
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        auth: state.auth,
        adminIsSuper: _.get(state.auth.admin, 'superAdmin', false),
        adminSeniorCenter: _.get(state.auth.admin, 'seniorCenter'),
        errors: state.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        registerActivity: (activityData, history) => dispatch(addActivity(activityData, history))
    };
};

AddActivity.propTypes = propTypes;
export default connect(
    mapStateToProps,
    mapDispatchToProps
)(AddActivity);
