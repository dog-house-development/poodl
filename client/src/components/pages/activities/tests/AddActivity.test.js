import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';

import { AddActivity, mapStateToProps, mapDispatchToProps } from '../AddActivity';

configure({ adapter: new Adapter() });

describe('Add activity tests', () => {
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                auth: {
                    isAuthenticated: true,
                    loading: false,
                    admin: {
                        superAdmin: false,
                        seniorCenter: 123,
                        firstName: 'Frog',
                        lastName: 'Anderson'
                    },
                    errors: {}
                },
                activities: {
                    errors: {}
                }
            },
            _state
        );
        wrapper = shallow(
            <AddActivity
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
        wrapper.setState({
            name: 'Test',
            description: 'test description',
            date: '1111-11-11T11:11:11',
            startTime: '1111-11-11T11:11:11',
            endTime: '1111-11-11T12:11:11',
            errors: { description: 'Description is required' }
        });
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
