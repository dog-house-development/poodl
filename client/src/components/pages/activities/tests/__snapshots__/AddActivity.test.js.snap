// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add activity tests render should render correctly 1`] = `
<div
  className="add-activity-container"
>
  <Link
    className="button small tertiary"
    replace={false}
    to="/activities"
  >
    <i
      className="material-icons"
    >
      keyboard_backspace
    </i>
     Back to all activities
  </Link>
  <div
    className="add-activity-panel panel"
  >
    <h1
      className="panel-title"
    >
      Add Activity
    </h1>
    <form
      noValidate={true}
      onSubmit={[Function]}
    >
      <OnClickOutside(ComboBox)
        eventTypes={
          Array [
            "mousedown",
            "touchstart",
          ]
        }
        excludeScrollbar={false}
        id="name"
        label="Name"
        onChange={[Function]}
        outsideClickIgnoreClass="ignore-react-onclickoutside"
        placeholder="Name"
        preventDefault={false}
        stopPropagation={false}
        value="Test"
      />
      <Field
        autoComplete="off"
        id="description"
        key="description"
        label="Description"
        onChange={[Function]}
        placeholder="Description"
        size="normal"
        type="text"
      />
      <div
        className="side-by-side"
      >
        <TimePicker
          ampmLabel=""
          hourLabel=""
          minuteLabel=""
          name="startTime"
          onChange={[Function]}
          time={"2019-03-25T05:00:00.000Z"}
          title="Start Time"
        />
        <TimePicker
          ampmLabel=""
          hourLabel=""
          minuteLabel=""
          name="endTime"
          onChange={[Function]}
          time={"2019-03-25T06:00:00.000Z"}
          title="End Time"
        />
      </div>
      <div
        className="side-by-side"
      >
        <DatePicker
          date="1111-11-11T11:11:11"
          dayLabel=""
          error=""
          monthLabel=""
          name="startDate"
          onChange={[Function]}
          title="Date"
          yearLabel=""
        />
      </div>
      <Button
        content="Multiple Days"
        dropdownButton={false}
        kind="secondary"
        onClick={[Function]}
        size="small"
        title="The activity spans multiple days"
        type="button"
      />
      <Button
        content="Add Activity"
        dropdownButton={false}
        formButton={true}
        kind="primary"
        size="medium"
        type="submit"
      />
    </form>
  </div>
</div>
`;
